sklearn/preprocessing/_encoders.py

```diff
@@ -290,9 +311,9 @@ def _identify_infrequent(self, category_count, n_samples, col_idx):
             infrequent_mask = np.zeros(category_count.shape[0], dtype=bool)
 
         n_current_features = category_count.size - infrequent_mask.sum() + 1
-        if self.max_categories is not None and self.max_categories < n_current_features:
+        if self._has_infrequent_categories(n_current_features, col_idx):
             # max_categories includes the one infrequent category
-            frequent_category_count = self.max_categories - 1
+            frequent_category_count = self._get_frequent_category_count(col_idx)
```

Comment:
Can you explain a bit why you moved these to separate methods? And why those methods take a `col_idx` argument that is unused?

I think my preference would be to leave this code unchanged, unless there is a reason related to this PR to change them.

Reply:
I moved these bits to separate methods so that I can override them in `OrdinalEncoder`. Specifically, since `max_categories` can also be an `array-like` or `dict`, I need to use `col_idx` to fetch the correct feature number. As this only applies for `OrdinalEncoder`, `col_idx` remains unused in the base class.